power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -170,21 +167,277 @@ template <bool sym> class NewtonRaphsonSESolver {
+                        auto const& injection = measured_value.bus_injection(row);
+                        auto const injection_jacobian = power_flow_jacobian_i(
+                            gs_minus_bc, gc_plus_bs, abs_ui_inv, partial_calculated_p, partial_calculated_q);
+                        add_injection_jacobian(block, injection_jacobian);
+
+                        rhs_block.tau_p() += injection.value.real();
+                        rhs_block.tau_q() += injection.value.imag();
+
+                        // TODO Confirm if correct
+                        block.r_P_theta() += RealTensor<sym>{
+                            RealValue<sym>{RealValue<sym>{1.0} / injection.p_variance / injection.p_variance}};
+                        block.r_Q_v() += RealTensor<sym>{
+                            RealValue<sym>{RealValue<sym>{1.0} / injection.q_variance / injection.q_variance}};
+                    } else {
+                        auto const injection_jacobian = power_flow_jacobian_j(gs_minus_bc, gc_plus_bs, abs_uj_inv);
+                        add_injection_jacobian(block, injection_jacobian);
+
+                        // subtract f(x) incrementally
+                        rhs_block.tau_p() -= partial_calculated_p;
+                        rhs_block.tau_q() -= partial_calculated_q;
+                    }
+
+                }
+                // injection measurement not exist
+                else {
+                    // Q_ij = 0
+                    // R_ii = -1.0, only diagonal
+                    // assign -1.0 to diagonal of 3x3 tensor, for asym
+                    if (row == col) {
+                        block.r_P_theta() = RealTensor<sym>{-1.0};
+                        block.r_Q_v() = RealTensor<sym>{-1.0};
+                    }
+                }
+
+                // Lagrange Multiplier: eta_i += sum_j (q_ji^T . phi_j)
+                rhs_block.eta_theta() +=
+                    dot(block.q_P_theta(), x_[col].phi_p()) + dot(block.q_Q_theta(), x_[col].phi_q());
+                rhs_block.eta_v() += dot(block.q_P_v(), x_[col].phi_p()) + dot(block.q_Q_v(), x_[col].phi_q());
+            }
+        }
 
+        // loop all transpose entry for QH
+        // assign the transpose of the transpose entry of Q
+        for (Idx data_idx_lu = 0; data_idx_lu != y_bus.nnz_lu(); ++data_idx_lu) {
+            // skip for fill-in
+            if (y_bus.map_lu_y_bus()[data_idx_lu] == -1) {
+                continue;
+            }
+            Idx const data_idx_tranpose = y_bus.lu_transpose_entry()[data_idx_lu];
+            data_gain_[data_idx_lu].qt_P_theta() = data_gain_[data_idx_tranpose].q_P_theta();
+            data_gain_[data_idx_lu].qt_P_v() = data_gain_[data_idx_tranpose].q_Q_theta();
+            data_gain_[data_idx_lu].qt_Q_theta() = data_gain_[data_idx_tranpose].q_P_v();
+            data_gain_[data_idx_lu].qt_Q_v() = data_gain_[data_idx_tranpose].q_Q_v();
+        }
```

Comment:
you can create a separate function for this, e.g. `make_symmetric_from_lower_triangle`